crates/sui-transaction-builder/src/lib.rs

```diff
@@ -0,0 +1,917 @@
+            return Err(anyhow!("Commands provided, but no inputs."));
+        }
+        if self.gas.is_empty() {
+            return Err(anyhow!("No gas objects provided"));
+        }
+
+        Ok(Transaction {
+            kind: sui_types::types::TransactionKind::ProgrammableTransaction(
+                sui_types::types::ProgrammableTransaction {
+                    inputs: self
+                        .inputs
+                        .iter()
+                        .map(|i| i.try_into())
+                        .collect::<Result<Vec<_>, _>>()
+                        .map_err(|e| anyhow!("Failed to convert inputs into InputArgument: {e}"))?,
+                    commands: self.commands.clone(),
+                },
+            ),
+            sender: self.sender,
+            gas_payment: {
+                GasPayment {
+                    objects: self
+                        .gas
+                        .clone()
+                        .into_iter()
+                        .map(|o| o.try_into())
+                        .collect::<Result<Vec<_>, _>>()
+                        .map_err(|_| anyhow!("Failed to convert gas objects into GasPayment"))?,
+                    owner: self.sponsor.unwrap_or(self.sender),
+                    price: self.gas_price,
+                    budget: self.gas_budget,
+                }
+            },
+            expiration: self.expiration.unwrap_or(TransactionExpiration::None),
+        })
+    }
+
+    /// Try to finish the transaction, but if not all parts are resolved, it will return an
+    /// [`UnresolvedTransaction`]. This can be used to resolve the transaction later by calling the
+    /// [`resolve_transaction`] method.
+    pub fn try_finish(&self) -> Result<TransactionResolution, Error> {
+        match self.finish() {
+            Ok(tx) => Ok(TransactionResolution::Resolved(tx)),
+            Err(_) => Ok(TransactionResolution::Unresolved(UnresolvedTransaction {
+                ptb: UnresolvedProgrammableTransaction {
+                    inputs: self.inputs.clone().into_iter().map(|x| x.into()).collect(),
+                    commands: self.commands.clone(),
+                },
+                sender: self.sender,
+                gas_payment: Some(UnresolvedGasPayment {
+                    objects: self
+                        .gas
+                        .clone()
+                        .into_iter()
+                        .map(to_unresolved_obj_ref)
+                        .collect(),
+                    owner: self.sponsor.unwrap_or(self.sender),
+                    price: Some(self.gas_price),
+                    budget: Some(self.gas_budget),
+                }),
+                expiration: self.expiration.unwrap_or(TransactionExpiration::None),
+            })),
+        }
+    }
+
+    /// Takes an unresolved transaction and attempts to resolve it by querying the GraphQL service.
+    pub fn resolve_transaction(
+        &self,
```

Comment:
This needs to take in a `client` as well, to query the RPC for resolving this transaction.